dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/shared/GridAdaptor.java

```diff
@@ -83,10 +97,72 @@ public Grid createGrid( final List<GridHeader> headers, final Map<Column, List<O
                 rowsAdded = true;
             }
 
-            // Note that the header column must match the result map key.
-            grid.addHeader( header ).addColumn( columnRows );
+            if ( !commonQueryRequest.isSkipHeaders() )
+            {
+                // Note that the header column must match the result map key.
+                grid.addHeader( header );
+            }
+
+            grid.addColumn( columnRows );
+        }
+
+        if ( !commonQueryRequest.isSkipMeta() )
+        {
+            grid.setMetaData( getMetadata( teiQueryParams, commonQueryRequest ) );
         }
 
         return grid;
     }
+
+    /**
+     * returns metadata based on teiQueryParams and rendered by
+     * commonQueryRequest
+     *
+     * @param teiQueryParams
+     * @param commonQueryRequest
+     * @return
+     */
+    private static Map<String, Object> getMetadata( final TeiQueryParams teiQueryParams,
+        final CommonQueryRequest commonQueryRequest )
+    {
+        final Map<String, Object> metadata = new HashMap<>();
+
+        metadata.put( PAGER.getKey(),
+            commonQueryRequest.isTotalPages()
+                ? new Pager( commonQueryRequest.getPage(), 1, commonQueryRequest.getPageSize() )
+                : new SlimPager( commonQueryRequest.getPage(), commonQueryRequest.getPageSize(), true ) );
+
+        if ( teiQueryParams.getCommonParams() == null )
+        {
+            return metadata;
+        }
+
+        if ( teiQueryParams.getCommonParams().getDimensionIdentifiers() != null )
+        {
+            final Map<String, MetadataItem> metadataItems = new HashMap<>();
+
+            teiQueryParams.getCommonParams().getDimensionIdentifiers()
+                .forEach(
+                    dimIdentifiers -> dimIdentifiers.forEach( di -> di.getDimension().getDimensionalObject().getItems()
+                        .forEach( dio -> metadataItems.put( dio.getUid(),
+                            commonQueryRequest.isIncludeMetadataDetails()
+                                ? new MetadataItem( dio.getDisplayName(), dio )
+                                : new MetadataItem( dio.getDisplayName() ) ) ) ) );
```

Comment:
we can take advantage of `flatMap()` here
```suggestion
            teiQueryParams.getCommonParams().getDimensionIdentifiers()
                .stream()
                .flatMap( Collection::stream )
                .map( DimensionIdentifier::getDimension )
                .map( DimensionParam::getDimensionalObject )
                .map( DimensionalObject::getItems )
                .flatMap( Collection::stream )
                .forEach( dio -> metadataItems.put( dio.getUid(),
                    commonQueryRequest.isIncludeMetadataDetails()
                        ? new MetadataItem( dio.getDisplayName(), dio )
                        : new MetadataItem( dio.getDisplayName() ) ) );
```